Library/Homebrew/cleanup.rb

```diff
@@ -519,5 +522,36 @@ def prune_prefix_symlinks_and_directories
       print "and #{d} directories " if d.positive?
       puts "from #{HOMEBREW_PREFIX}"
     end
+
+    def self.autoremove(dry_run: false)
+      require "cask/caskroom"
+
+      # If this runs after install, uninstall, reinstall or upgrade,
+      # the cache of installed formulae may no longer be valid.
+      Formula.clear_cache unless dry_run
```

Comment:
I wonder if clearing the cache should be handled by `install`, `uninstall`, etc, rather than have to be managed by callers.

Reply:
That would be easy to add to `uninstall` because `autoremove` only runs once after all the uninstalls have finished. 

But with `install`, `reinstall` and `upgrade` it's a little more difficult. In those cases, `autoremove` occasionally gets run after the periodic cleanup gets triggered by the first call to `Cleanup#cleanup_install_formula!`. This means that `autoremove` gets run in the middle of another cache changing set of actions which is not predictable.

If as suggested in #13569, we decide to run the periodic cleanup before those commands the cache won't be invalidated and the cache clearing would only be necessary for `brew uninstall`.